src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingNullCaseInSwitchCheck.java

```diff
@@ -0,0 +1,205 @@
+ * As of Java 21,
+ * it is now possible to integrate a null check within the switch,
+ * eliminating the risk of {@code NullPointerException} and simplifies the code
+ * as there is no need for an external null check before entering the switch.
+ * </p>
+ * <p>
+ * Limitations due to Checkstyle not being type-aware:
+ * </p>
+ * <ul>
+ * <li>
+ * This check does not support enums; syntactically, these are no different from other constants.
+ * </li>
+ * </ul>
+ * <p>
+ * See the <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-15.html#jls-15.28">
+ * Java Language Specification</a> for more information about switch statements and expressions.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code missing.switch.nullcase}
+ * </li>
+ * </ul>
+ *
+ * @since 10.18.0
+ */
+
+@StatelessCheck
+public class MissingNullCaseInSwitchCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "missing.switch.nullcase";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.LITERAL_SWITCH};
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final List<DetailAST> caseLabels = getAllCaseLabels(ast);
+        final boolean hasNullCaseLabel = caseLabels.stream()
+                .anyMatch(MissingNullCaseInSwitchCheck::hasLiteralNull);
+        if (!hasNullCaseLabel) {
+            final boolean hasPatternCaseLabel = caseLabels.stream()
+                .anyMatch(MissingNullCaseInSwitchCheck::hasPatternCaseLabel);
+            final boolean hasStringCaseLabel = caseLabels.stream()
+                .anyMatch(MissingNullCaseInSwitchCheck::hasStringCaseLabel);
+            if (hasPatternCaseLabel || hasStringCaseLabel) {
+                log(ast, MSG_KEY);
+            }
+        }
+    }
+
+    /**
+     * Gets all case labels in the given switch AST node.
+     *
+     * @param detailAST the AST node representing {@code LITERAL_SWITCH}
+     * @return a list of all case labels in the switch
+     */
+    private static List<DetailAST> getAllCaseLabels(DetailAST detailAST) {
+        final List<DetailAST> caseLabels = new ArrayList<>();
+        DetailAST ast = detailAST.getFirstChild();
+        while (ast != null) {
+            // case group token may have several LITERAL_CASE tokens
+            TokenUtil.forEachChild(ast, TokenTypes.LITERAL_CASE, caseLabels::add);
+            ast = ast.getNextSibling();
+        }
+        return Collections.unmodifiableList(caseLabels);
```

Comment:
Did something require this to unmodifable? It seems odd to do this for a list that will never leave this class.

Reply:
Data should always be final/unmodifiable unless we need to modify it.

Reply:
Collections should be unmodifiable unless we need to modify them on purpose. This is a common best practice we use it a lot in our codebase with collections.